File: hackerrank/problem solving/algorithms/2 - implementation/7 - divisible sum pairs/divisible_sum_pairs.py
# https://www.hackerrank.com/challenges/divisible-sum-pairs/problem?isFullScreen=true

def main():
    _, k = read_numbers()
    numbers = read_numbers()
    print(divisible_sum_pairs(numbers, k))


def read_numbers():
    return list(map(int, input().split()))


# n: length of numbers
# k: length of freq
# T: O(n + k)
# S: O(k) extra space
def divisible_sum_pairs(numbers, k):
    freq = init_remainder_frequency(numbers, k)
    return (count_pairs_with_remainder_0(freq) +
            count_complementary_remainder_pairs(freq) +
            count_pairs_with_remainder_k_half(freq))


def init_remainder_frequency(numbers, k):
    freq = [0] * k
    for x in numbers:
        freq[x % k] += 1
    return freq


def count_pairs_with_remainder_0(freq):
    return pair_count(freq[0])


def pair_count(n):
    return n * (n - 1) // 2


def count_complementary_remainder_pairs(freq):
    return sum(freq[i] * freq[-i] for i in range(1, (len(freq) + 1) // 2))


def count_pairs_with_remainder_k_half(freq):
    n = len(freq)
    return pair_count(freq[n // 2]) if n % 2 == 0 else 0


if __name__ == '__main__':
    main()

File: hackerrank/problem solving/algorithms/2 - implementation/7 - divisible sum pairs/divisible_sum_pairs.js
// https://www.hackerrank.com/challenges/divisible-sum-pairs/problem?isFullScreen=true

function main() {
  const [, k] = readNumbers();
  const numbers = readNumbers();
  console.log(divisibleSumPairs(numbers, k));
}

function readNumbers() {
  return readLine()
    .split(' ')
    .map((x) => +x);
}

// n: length of numbers
// k: length of freq
// T: O(n + k)
// S: O(k) extra space
function divisibleSumPairs(numbers, k) {
  const freq = initRemainderFrequency(numbers, k);
  return (
    countPairsWithRemainder0(freq) +
    countComplementaryRemainderPairs(freq) +
    countPairsWithRemainderKHalf(freq)
  );
}

function initRemainderFrequency(numbers, k) {
  const freq = new Array(k).fill(0);
  for (const x of numbers) freq[x % k]++;
  return freq;
}

function countPairsWithRemainder0(freq) {
  return pairCount(freq[0]);
}

function pairCount(n) {
  return (n * (n - 1)) / 2;
}

function countComplementaryRemainderPairs(freq) {
  let count = 0;
  for (let i = 1, n = Math.trunc((freq.length + 1) / 2); i < n; i++)
    count += freq[i] * freq.at(-i);
  return count;
}

function countPairsWithRemainderKHalf(freq) {
  const n = freq.length;
  return n % 2 === 0 ? pairCount(freq[n / 2]) : 0;
}

////////////////////////////////////////////////

const readline = require('readline');

const rl = readline.createInterface({ input: process.stdin });
let inputLines = [];
rl.on('line', (line) => inputLines.push(line));
rl.on('close', main);

function readLine() {
  return inputLines.shift();
}
